File: python/x402/mechanisms/svm/default_assets.py
"""Default USD-pegged assets for SVM payment schemes.

Index 0 of each network list is the ``"$0.10"`` default. Extra entries enable
suffixed prices such as ``"1 USDT"`` via core ``parse_money``.
"""

from __future__ import annotations

from typing import TypedDict

from .constants import (
    SOLANA_DEVNET_CAIP2,
    SOLANA_MAINNET_CAIP2,
    SOLANA_TESTNET_CAIP2,
    TOKEN_2022_PROGRAM_ADDRESS,
    TOKEN_PROGRAM_ADDRESS,
    USDC_DEVNET_ADDRESS,
    USDC_MAINNET_ADDRESS,
    USDC_TESTNET_ADDRESS,
    V1_TO_V2_NETWORK_MAP,
)


class SvmDefaultAsset(TypedDict):
    """Default stablecoin entry for SVM exact scheme."""

    asset: str
    decimals: int
    symbol: str
    token_program: str


DEFAULT_ASSETS: dict[str, list[SvmDefaultAsset]] = {
    SOLANA_MAINNET_CAIP2: [
        {
            "asset": USDC_MAINNET_ADDRESS,
            "decimals": 6,
            "symbol": "USDC",
            "token_program": TOKEN_PROGRAM_ADDRESS,
        },
        {
            "asset": "Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB",
            "decimals": 6,
            "symbol": "USDT",
            "token_program": TOKEN_PROGRAM_ADDRESS,
        },
        {
            "asset": "2u1tszSeqZ3qBWF3uNGPFc8TzMk2tdiwknnRMWGWjGWH",
            "decimals": 6,
            "symbol": "USDG",
            "token_program": TOKEN_2022_PROGRAM_ADDRESS,
        },
        {
            "asset": "2b1kV6DkPAnxd5ixfnxCpjxmKwqjjaYmCZfHsFu24GXo",
            "decimals": 6,
            "symbol": "PYUSD",
            "token_program": TOKEN_2022_PROGRAM_ADDRESS,
        },
        {
            "asset": "CASHx9KJUStyftLFWGvEVf59SGeG9sh5FfcnZMVPCASH",
            "decimals": 6,
            "symbol": "CASH",
            "token_program": TOKEN_2022_PROGRAM_ADDRESS,
        },
    ],
    SOLANA_DEVNET_CAIP2: [
        {
            "asset": USDC_DEVNET_ADDRESS,
            "decimals": 6,
            "symbol": "USDC",
            "token_program": TOKEN_PROGRAM_ADDRESS,
        },
        {
            "asset": "4F6PM96JJxngmHnZLBh9n58RH4aTVNWvDs2nuwrT5BP7",
            "decimals": 6,
            "symbol": "USDG",
            "token_program": TOKEN_2022_PROGRAM_ADDRESS,
        },
        {
            "asset": "CXk2AMBfi3TwaEL2468s6zP8xq9NxTXjp9gjMgzeUynM",
            "decimals": 6,
            "symbol": "PYUSD",
            "token_program": TOKEN_2022_PROGRAM_ADDRESS,
        },
    ],
    SOLANA_TESTNET_CAIP2: [
        {
            "asset": USDC_TESTNET_ADDRESS,
            "decimals": 6,
            "symbol": "USDC",
            "token_program": TOKEN_PROGRAM_ADDRESS,
        },
        {
            "asset": "4F6PM96JJxngmHnZLBh9n58RH4aTVNWvDs2nuwrT5BP7",
            "decimals": 6,
            "symbol": "USDG",
            "token_program": TOKEN_2022_PROGRAM_ADDRESS,
        },
        {
            "asset": "CXk2AMBfi3TwaEL2468s6zP8xq9NxTXjp9gjMgzeUynM",
            "decimals": 6,
            "symbol": "PYUSD",
            "token_program": TOKEN_2022_PROGRAM_ADDRESS,
        },
    ],
}


def _resolve_network_key(network: str) -> str:
    """Map CAIP-2 or v1 name to a ``DEFAULT_ASSETS`` key (mirrors ``normalize_network``)."""
    if ":" in network:
        supported = [SOLANA_MAINNET_CAIP2, SOLANA_DEVNET_CAIP2, SOLANA_TESTNET_CAIP2]
        if network not in supported:
            raise ValueError(f"Unsupported SVM network: {network}")
        return network

    caip2_network = V1_TO_V2_NETWORK_MAP.get(network)
    if not caip2_network:
        raise ValueError(f"Unsupported SVM network: {network}")
    return caip2_network


def get_default_asset(network: str, symbol: str | None = None) -> SvmDefaultAsset:
    """Look up a default asset by network and optional ticker.

    Args:
        network: CAIP-2 or v1 network.
        symbol: Ticker; omit for the network default.

    Returns:
        Matching entry.

    Raises:
        ValueError: If network or ticker is unknown.
    """
    key = _resolve_network_key(network)
    assets = DEFAULT_ASSETS.get(key)
    if not assets:
        raise ValueError(f"No default asset configured for network {network}")
    if not symbol:
        return assets[0]
    normalized = symbol.upper()
    for entry in assets:
        if entry["symbol"].upper() == normalized:
            return entry
    raise ValueError(f"No {symbol} default asset configured for network {network}")


def find_default_asset(asset: str, network: str) -> SvmDefaultAsset | None:
    """Reverse lookup by mint address and network.

    Args:
        asset: Mint address from payment requirements.
        network: CAIP-2 or v1 network.

    Returns:
        Matching entry, or ``None``.
    """
    key = _resolve_network_key(network)
    assets = DEFAULT_ASSETS.get(key)
    if not assets:
        return None
    for entry in assets:
        if entry["asset"] == asset:
            return entry
    return None

File: typescript/packages/mechanisms/svm/src/defaultAssets.ts
import type {
  DefaultAsset,
  DefaultAssetTable,
  FindDefaultAsset,
  GetDefaultAsset,
  Network,
} from "@x402/core/types";
import {
  SOLANA_DEVNET_CAIP2,
  SOLANA_MAINNET_CAIP2,
  SOLANA_TESTNET_CAIP2,
  TOKEN_2022_PROGRAM_ADDRESS,
  TOKEN_PROGRAM_ADDRESS,
} from "./constants";
import { normalizeNetwork } from "./constants";

export type SvmDefaultAsset = DefaultAsset & {
  /** Program owning the mint: SPL Token or Token-2022. */
  tokenProgram: string;
};

export const USDC_MAINNET_ADDRESS = "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v";
export const USDC_DEVNET_ADDRESS = "4zMMC9srt5Ri5X14GAgXhaHii3GnPAEERYPJgZJDncDU";
export const USDC_TESTNET_ADDRESS = "4zMMC9srt5Ri5X14GAgXhaHii3GnPAEERYPJgZJDncDU";

/** Default USD-pegged assets by CAIP-2 network; index 0 is the `"$0.10"` default. */
export const DEFAULT_ASSETS: DefaultAssetTable<SvmDefaultAsset> = {
  [SOLANA_MAINNET_CAIP2]: [
    {
      asset: USDC_MAINNET_ADDRESS,
      decimals: 6,
      symbol: "USDC",
      tokenProgram: TOKEN_PROGRAM_ADDRESS,
    },
    {
      asset: "Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB",
      decimals: 6,
      symbol: "USDT",
      tokenProgram: TOKEN_PROGRAM_ADDRESS,
    },
    {
      asset: "2u1tszSeqZ3qBWF3uNGPFc8TzMk2tdiwknnRMWGWjGWH",
      decimals: 6,
      symbol: "USDG",
      tokenProgram: TOKEN_2022_PROGRAM_ADDRESS,
    },
    {
      asset: "2b1kV6DkPAnxd5ixfnxCpjxmKwqjjaYmCZfHsFu24GXo",
      decimals: 6,
      symbol: "PYUSD",
      tokenProgram: TOKEN_2022_PROGRAM_ADDRESS,
    },
    {
      asset: "CASHx9KJUStyftLFWGvEVf59SGeG9sh5FfcnZMVPCASH",
      decimals: 6,
      symbol: "CASH",
      tokenProgram: TOKEN_2022_PROGRAM_ADDRESS,
    },
  ],
  [SOLANA_DEVNET_CAIP2]: [
    {
      asset: USDC_DEVNET_ADDRESS,
      decimals: 6,
      symbol: "USDC",
      tokenProgram: TOKEN_PROGRAM_ADDRESS,
    },
    {
      asset: "4F6PM96JJxngmHnZLBh9n58RH4aTVNWvDs2nuwrT5BP7",
      decimals: 6,
      symbol: "USDG",
      tokenProgram: TOKEN_2022_PROGRAM_ADDRESS,
    },
    {
      asset: "CXk2AMBfi3TwaEL2468s6zP8xq9NxTXjp9gjMgzeUynM",
      decimals: 6,
      symbol: "PYUSD",
      tokenProgram: TOKEN_2022_PROGRAM_ADDRESS,
    },
  ],
  [SOLANA_TESTNET_CAIP2]: [
    {
      asset: USDC_TESTNET_ADDRESS,
      decimals: 6,
      symbol: "USDC",
      tokenProgram: TOKEN_PROGRAM_ADDRESS,
    },
    {
      asset: "4F6PM96JJxngmHnZLBh9n58RH4aTVNWvDs2nuwrT5BP7",
      decimals: 6,
      symbol: "USDG",
      tokenProgram: TOKEN_2022_PROGRAM_ADDRESS,
    },
    {
      asset: "CXk2AMBfi3TwaEL2468s6zP8xq9NxTXjp9gjMgzeUynM",
      decimals: 6,
      symbol: "PYUSD",
      tokenProgram: TOKEN_2022_PROGRAM_ADDRESS,
    },
  ],
};

/**
 * Map CAIP-2 or v1 name to a {@link DEFAULT_ASSETS} key.
 *
 * @param network - CAIP-2 or legacy Solana network id
 * @returns Normalized CAIP-2 network key
 */
function resolveNetworkKey(network: Network): string {
  return normalizeNetwork(network);
}

/**
 * Look up a default asset by network and optional ticker.
 *
 * @param network - CAIP-2 or v1 network
 * @param symbol - Ticker; omit for the network default
 * @returns Matching entry
 * @throws If network or ticker is unknown
 */
export const getDefaultAsset: GetDefaultAsset<SvmDefaultAsset> = (network, symbol?) => {
  const key = resolveNetworkKey(network);
  const assets = DEFAULT_ASSETS[key];
  if (!assets || assets.length === 0) {
    throw new Error(`No default asset configured for network ${network}`);
  }
  if (!symbol) {
    return assets[0];
  }
  const normalized = symbol.toUpperCase();
  const match = assets.find(entry => entry.symbol.toUpperCase() === normalized);
  if (!match) {
    throw new Error(`No ${symbol} default asset configured for network ${network}`);
  }
  return match;
};

/**
 * Reverse lookup by asset id and network.
 *
 * @param asset - Mint address from payment requirements
 * @param network - CAIP-2 or v1 network
 * @returns Matching entry, or undefined
 */
export const findDefaultAsset: FindDefaultAsset<SvmDefaultAsset> = (asset, network) => {
  const key = resolveNetworkKey(network);
  const assets = DEFAULT_ASSETS[key];
  if (!assets) {
    return undefined;
  }
  return assets.find(entry => entry.asset === asset);
};
